xdem/volume.py

```diff
@@ -0,0 +1,194 @@
+from __future__ import annotations
+
+import datetime
+from typing import Any, Optional, Union
+
+import geoutils as gu
+import numpy as np
+import pandas as pd
+
+import xdem
+
+
+class dDEM(gu.georaster.Raster):   # pylint: disable=invalid-name
+    """A difference-DEM object."""
+
+    def __init__(self, raster: gu.georaster.Raster, start_time: np.datetime64, end_time: np.datetime64,
+                 error: Optional[Any] = None):
+        """
+        Create a dDEM object from a Raster.
+
+        :param raster: A georeferenced Raster object.
+        :param start_time: The starting time of the dDEM.
+        :param end_time: The end time of the dDEM.
+        :param error: An error measure for the dDEM (UNUSED).
+
+        :returns: A new dDEM instance.
+        """
+
+        self.__dict__ = raster.__dict__
+        self.start_time = start_time
+        self.end_time = end_time
+        self.error = error
+
+    def __str__(self) -> str:
+        """Return a summary of the dDEM."""
+        return f"dDEM from {self.start_time} to {self.end_time}.\n\n{super().__str__()}"
+
+    @property
+    def time(self) -> np.timedelta64:
+        """Get the time duration."""
+        return self.end_time - self.start_time
+
+    def from_array(data: np.ndarray, transform, crs, start_time, end_time, error=None, nodata=None) -> dDEM:
+        """
+        Create a new dDEM object from an array.
+
+        :param data: The dDEM data array.
+        :param transform: A geometric transform.
+        :param crs: The coordinate reference system of the dDEM.
+        :param start_time: The starting time of the dDEM.
+        :param end_time: The end time of the dDEM.
+        :param error: An error measure for the dDEM.
+        :param nodata: The nodata value.
+
+        :returns: A new dDEM instance.
+        """
+        return dDEM(
+            gu.georaster.Raster.from_array(
+                data=data,
+                transform=transform,
+                crs=crs,
+                nodata=nodata
+            ),
+            start_time=start_time,
+            end_time=end_time,
+            error=error,
+        )
+
+
+class tDEM:
+    """A temporal collection of DEMs."""
+
+    def __init__(self, dems: list[gu.georaster.Raster], timestamps: list[datetime.datetime],
```

Comment:
Preferably a list of xdem.dem.DEM objects. Basically the same object, but possible timestamp/metadata/vref attached

Reply:
Totally agree and I just completely forgot that. I'll fix it. I still think it would be nice to accept Raster objects, but then the `timestamps` argument is needed.

Reply:
Wait I misunderstood. The DEM class doesn't have a timestamp attached right now. Scratch the optionality of the `timestamps` argument.

Reply:
The DEM class inherits from the SatImg class which has a .datetime attribute! See my general comment on passing the datetimes directly to xarray.

Reply:
@rhugonnet I changed the default tDEM collection to be of `DEM` classes in 4538bee. It still takes `Raster`s, but will convert them in the `__init__()`

I just saw your datetime comment. I'll experiment with that now.

Reply:
@rhugonnet I added `.datetime` attribute parsing in df80f81.

The tDEM instantiation is now extremely simple, assuming the datetime attribute is set:
```python

tdem = tDEM([dem_2009, dem_1990])  # The first DEM in the list is parsed as the reference

```

Reply:
Those classes are gonna be so neat =D !